### PR TITLE
Cmake massive alteration

### PR DESCRIPTION
**Pull request by cybik**
Massive modifications introducing fixes and reintroducing a feature.

- Paving the way for cross-compile using cmake alternate toolchains, such as for Android
- Get CrashManager back into Linux
- Got rid of the generation of a binary blob generation that was ultimately incompatible with cross-compiling, instead recursively calling cmake from cmake.

Small-ish issue: due to the use of one specific "new" feature that got introduced later, and because the Android toolchain is more stable with 3.0.99+, this also introduces the need for users to install cmake 3.1 if they are to compile the Android part.

### CODE REVIEW

**Comment by cybik** [comment]
For those who'll ask, the Android part is still not even CLOSE.

File: CMake/DefineOptions.cmake [hunk 1]
```diff
@@ -28,19 +25,24 @@ option(WITH_GPL_LIBS "Build with GPL libraries." ON)
...
     option(WITH_PARALLEL_PORT "Build with Parallel Lights I/O Support." OFF)
     option(WITH_CRASH_HANDLER "Build with Crash Handler Support." ON)
-  endif()
 endif()
+
+if(NOT ANDROID)
+  # This option handles if we use SSE2 processing.
+  option(WITH_SSE2 "Build with SSE2 Optimizations." ON)
+elseif(ANDROID)
+  option(WITH_CRASH_HANDLER "Build with Crash Handler Support." ON)
```
Comment (wolfman2000) [inline CMake/DefineOptions.cmake]: Considering that Linux was working just fine without the crash handler (except when it crashes), should this perhaps be an option for all but Windows instead?

File: CMake/DefineOptions.cmake [hunk 2]
```diff
@@ -28,19 +25,24 @@ option(WITH_GPL_LIBS "Build with GPL libraries." ON)
...
-  if(LINUX)
+elseif(LINUX)
     option(WITH_FFMPEG "Build with FFMPEG." ON)
     # Builder beware: later versions of ffmpeg may break!
     option(WITH_SYSTEM_FFMPEG "Build with the system's FFMPEG." OFF)
     option(WITH_CRYSTALHD_DISABLED "Build FFMPEG without Crystal HD support." OFF)
     option(WITH_TTY "Build with Linux TTY Input Support." OFF)
     option(WITH_PROFILING "Build with Profiling Support." OFF)
-    option(WITH_GLES2 "Build with OpenGL ES 2.0 Support." ON)
+    #option(WITH_GLES2 "Build with OpenGL ES 2.0 Support." OFF)
```
Comment (wolfman2000) [inline CMake/DefineOptions.cmake]: What is the status of this GLES option now?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -1,14 +1,30 @@
...
+if(NOT ANDROID)
 # Version Updater library.
 add_subdirectory(version_updater)
 
 # This part may move to a different file: it's the SM Data project.
 add_custom_command(
-  OUTPUT ${SM_SRC_DIR}/verstub.cpp
-  COMMAND VersionUpdater ${SM_SRC_DIR}/verstub.cpp
+  OUTPUT ${SM_SRC_DIR}/generated/verstub.cpp
+  COMMAND VersionUpdater ${SM_SRC_DIR}/generated/verstub.cpp
```
Comment (wolfman2000) [inline src/CMakeLists.txt]: I thought part of the idea was that there would not be a version updater program.

File: src/CMakeLists.txt [hunk 4]
```diff
@@ -212,9 +235,11 @@ if(MSVC)
   set(SM_COMPILE_FLAGS "${SM_COMPILE_FLAGS} /MP2")
 endif()
 
+if(${SM_COMPILE_FLAGS})
```
Comment (wolfman2000) [inline src/CMakeLists.txt]: Alright, that's a fair if block.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -342,12 +367,30 @@ else() # Linux
     RUNTIME_OUTPUT_DIRECTORY_RELWITHDEBINFO "${SM_ROOT_DIR}"
   )
 
+  if(${WITH_CRASH_HANDLER})
+    sm_add_compile_definition("${SM_EXE_NAME}" CRASH_HANDLER)
+    if(LINUX)
+      # used only in the ArchHooks_Unix.cpp file
```
Comment (wolfman2000) [inline src/CMakeLists.txt]: If this is only used with the `ArchHooks_Unix.cpp` file, then wouldn't it make more sense to add these preprocessor definitions only to the individual source files instead?

File: src/version_updater/version.hpp.in [hunk 6]
```diff
@@ -12,4 +12,6 @@ char const * const smVersionFull = "@SM_VERSION_FULL@";
 #endif
 char const * const smVersionTraditional = "@SM_VERSION_TRADITIONAL@";
 
+extern char const * const smVersionDate = "@SM_TIMESTAMP_DATE@";
```
Comment (wolfman2000) [inline src/version_updater/version.hpp.in]: Is `version.hpp.in` actually being used anymore? If not, I would advise removing it.

File: CMake/VerStubUtil.cmake [hunk 7]
```diff
@@ -0,0 +1,8 @@
+include("${CMAKE_CURRENT_LIST_DIR}/SMDefs.cmake")
+
+string(TIMESTAMP SM_TIMESTAMP_DATE "%Y%m%d")
+string(TIMESTAMP SM_TIMESTAMP_TIME "%H:%M:%S" UTC)
```
Comment (wolfman2000) [inline CMake/VerStubUtil.cmake]: Wasn't it already established that this would require cmake 3 or higher? This is a deal breaker.

**Comment by wolfman2000** [comment]
There is some potential here, don't get me wrong. However, I don't think all of the changes made here are acceptable.

I'll see if I can come up with something.

File: extern/CMakeProject-jpeg.cmake [hunk 8]
```diff
@@ -63,4 +63,6 @@ set_property(TARGET "jpeg" PROPERTY FOLDER "External Libraries")
 
 if(MSVC)
   sm_add_compile_definition("jpeg" _CRT_SECURE_NO_WARNINGS)
+elseif(ANDROID)
+  sm_add_compile_definition("jpeg" STDC_HEADERS=1)
```
Comment (wolfman2000) [inline extern/CMakeProject-jpeg.cmake]: That seems like a definition that maybe all builds should use.

**Comment by cybik** [comment]
Closing in favor of separated pull requests

**Pull request closed without merging**